File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/domain/AbstractPersistable.java
package com.coupang.springframework.data.requery.domain;

import com.coupang.kotlinx.objectx.AbstractValueObject;
import com.coupang.kotlinx.objectx.ToStringBuilder;
import io.requery.Persistable;
import io.requery.Superclass;
import io.requery.Transient;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

/**
 * AbstractPersistable
 *
 * @author debop@coupang.com
 * @since 18. 6. 4
 */
@Superclass
public abstract class AbstractPersistable<ID> extends AbstractValueObject implements Persistable {

    @Nullable
    abstract public ID getId();

    @io.requery.Transient
    public boolean isNew() {
        return getId() == null;
    }

    @Override
    public boolean equals(@Nullable Object other) {
        if (other == null)
            return false;

        if (other instanceof AbstractPersistable) {
            AbstractPersistable that = (AbstractPersistable) other;
            return (isNew() && that.isNew())
                   ? hashCode() == other.hashCode()
                   : Objects.equals(getId(), that.getId());

        }

        return false;
    }

    @Override
    public int hashCode() {
        return (getId() != null) ? getId().hashCode() : System.identityHashCode(this);
    }

    @NotNull
    @Override
    @Transient
    protected ToStringBuilder buildStringHelper() {
        return super.buildStringHelper()
            .add("id", getId());
    }

    private static final long serialVersionUID = 5460350519810267858L;
}

File: spring-data-requery-kotlin/src/main/kotlin/com/coupang/springframework/data/requery/domain/AbstractPersistable.kt
package com.coupang.springframework.data.requery.domain

import com.coupang.kotlinx.objectx.AbstractValueObject
import com.coupang.kotlinx.objectx.ToStringBuilder
import io.requery.Superclass
import io.requery.Transient
import java.util.*

/**
 * Requery Entity의 최상위 추상화 클래스입니다.
 *
 * 이 클래스를 상속받은 클래스는 반드시 hashCode() 를 재정의해주셔야 합니다.
 *
 * @author debop
 * @since 18. 5. 23
 */
@Superclass
abstract class AbstractPersistable<ID>: AbstractValueObject(), io.requery.Persistable {

    abstract val id: ID?

    @io.requery.Transient
    fun isNew(): Boolean = id == null

    override fun equals(other: Any?): Boolean {
        if(this === other) return true

        return when(other) {
            is AbstractPersistable<*> ->
                if(isNew() && other.isNew()) hashCode() == other.hashCode()
                else Objects.equals(id, other.id)
            else                      -> false
        }
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: System.identityHashCode(this)
    }

    @Transient
    override fun buildStringHelper(): ToStringBuilder {
        return super.buildStringHelper()
            .add("id", id)
    }
}